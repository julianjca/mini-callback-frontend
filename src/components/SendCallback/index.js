import React, { useState } from 'react'
import axios from 'axios'
import {
  Center,
  Container,
  Heading,
  Stack,
  Input,
  FormControl,
  FormLabel,
  Button,
  Select,
  useDisclosure,
  Modal,
  ModalOverlay,
  ModalContent,
  ModalHeader,
  ModalFooter,
  ModalCloseButton,
} from '@chakra-ui/react'
import cookie from 'react-cookies'
import { useRouter } from 'next/router'

const SuccessModal = ({ isOpen, onClose }) => (
  <Modal isOpen={isOpen} onClose={onClose}>
    <ModalOverlay />
    <ModalContent>
      <ModalHeader>Success!</ModalHeader>
      <ModalCloseButton />

      <ModalFooter>
        <Button colorScheme="blue" mr={3} onClick={onClose}>
          Close
        </Button>
      </ModalFooter>
    </ModalContent>
  </Modal>
)

const SendCallback = ({ businesses }) => {
  const router = useRouter()
  const [virtualAccount, setVirtualAccount] = useState('')
  const [bankCode, setBankCode] = useState('')
  const [transactionId, setTransactionId] = useState('')
  const [businessId, setBusinessId] = useState('')
  const { isOpen, onOpen, onClose } = useDisclosure()

  const resetForm = () => {
    setBusinessId('')
    setTransactionId('')
    setBankCode('')
    setVirtualAccount('')
  }

  const handleSubmit = async () => {
    await axios.post(
      'http://localhost:3030/callbacks',
      {
        virtual_account: virtualAccount,
        bank_code: bankCode,
        timestamp: new Date(),
        transaction_id: transactionId,
        business_id: businessId,
      },
      {
        headers: {
          Authorization: `Bearer ${cookie.load('userToken')}`,
        },
      },
    )
    onOpen()
    resetForm()
  }

  return (
    <>
      <SuccessModal isOpen={isOpen} onClose={onClose} />
      <Container pt={20} pb={20}>
        <Center>
          <Heading>Send callback</Heading>
        </Center>
        <Center mt={10}>
          <Stack spacing={5}>
            <FormControl id="va">
              <FormLabel>Virtual account</FormLabel>
              <Input
                value={virtualAccount}
                width={300}
                type="text"
                placeholder="Virtual Account"
                size="md"
                onChange={e => setVirtualAccount(e.target.value)}
                data-testid="va-input"
              />
            </FormControl>
            <FormControl id="bank-code">
              <FormLabel>Bank code</FormLabel>
              <Input
                value={bankCode}
                width={300}
                type="text"
                placeholder="Bank code"
                size="md"
                onChange={e => setBankCode(e.target.value)}
                data-testid="bank-code-input"
              />
            </FormControl>
            <FormControl id="trx">
              <FormLabel>Transaction ID</FormLabel>
              <Input
                value={transactionId}
                width={300}
                type="text"
                placeholder="Bank code"
                size="md"
                onChange={e => setTransactionId(e.target.value)}
                data-testid="trx-id-input"
              />
            </FormControl>
            <FormControl id="business">
              <FormLabel>Business</FormLabel>
              <Select
                value={businessId}
                onChange={e => setBusinessId(e.target.value)}
                placeholder="Select option"
                data-testid="business-input"
              >
                {businesses.map(business => (
                  <option key={business.id} value={business.id}>
                    {business.businessName}
                  </option>
                ))}
              </Select>
            </FormControl>
            <Button
              onClick={handleSubmit}
              size="md"
              width={300}
              colorScheme="blue"
              data-testid="send-callback-button"
            >
              Send callback
            </Button>
            <Button
              variant="outline"
              onClick={() => router.push('/dashboard')}
              size="md"
              width={300}
              colorScheme="blue"
            >
              Go back
            </Button>
          </Stack>
        </Center>
      </Container>
    </>
  )
}

export default SendCallback
