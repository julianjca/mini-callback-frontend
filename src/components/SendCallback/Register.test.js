/* eslint-disable testing-library/prefer-screen-queries */
import React from 'react'
import { render, fireEvent, waitFor } from '../../../test-utils'
import axios from 'axios'

import SendCallback from './index'

jest.mock('axios')

describe('Register', () => {
  test('renders the right text', () => {
    const element = render(<SendCallback businesses={[]} />)

    expect(element).toMatchSnapshot()
  })

  test('renders email input, name input, and password input', () => {
    const element = render(<SendCallback businesses={[]} />)

    expect(element.getByTestId('va-input')).toBeInTheDocument()
    expect(element.getByTestId('trx-id-input')).toBeInTheDocument()
    expect(element.getByTestId('business-input')).toBeInTheDocument()
    expect(element.getByTestId('bank-code-input')).toBeInTheDocument()
  })

  test('runs register function', async () => {
    const element = render(<SendCallback businesses={[]} />)

    const axiosMock = jest.spyOn(axios, 'post').mockImplementation(() => Promise.resolve())

    const vaInput = element.getByTestId('va-input')
    const trxIdInput = element.getByTestId('trx-id-input')
    const businessInput = element.getByTestId('business-input')
    const bankCodeInput = element.getByTestId('bank-code-input')

    fireEvent.change(vaInput, { target: { value: 'VA' } })
    fireEvent.change(trxIdInput, { target: { value: 'TRX' } })
    fireEvent.change(businessInput, { target: { value: 'businessId' } })
    fireEvent.change(bankCodeInput, { target: { value: 'BCA' } })

    const button = element.getByTestId('send-callback-button')
    fireEvent.click(button)

    await waitFor(() => expect(axiosMock).toHaveBeenCalledTimes(1))

    axiosMock.mockRestore()
  })
})
