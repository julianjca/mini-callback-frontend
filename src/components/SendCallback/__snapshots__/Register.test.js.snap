// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register renders the right text 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-light"
  >
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto; top: 0px; right: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; top: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; top: 0px; right: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; bottom: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto; bottom: 0px; right: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; bottom: 0px; right: 0px;"
      />
    </div>
    <div>
      <div
        class="chakra-container css-kvcvn0"
      >
        <div
          class="css-gmuwbf"
        >
          <h2
            class="chakra-heading css-zey6tx"
          >
            Send callback
          </h2>
        </div>
        <div
          class="css-rwokqt"
        >
          <div
            class="chakra-stack css-ynqzo5"
          >
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="chakra-form__label css-1qqx8g3"
                for="va"
                id="va-label"
              >
                Virtual account
              </label>
              <input
                class="chakra-input css-1hv7q3p"
                data-testid="va-input"
                id="va"
                placeholder="Virtual Account"
                type="text"
                value=""
              />
            </div>
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="chakra-form__label css-1qqx8g3"
                for="bank-code"
                id="bank-code-label"
              >
                Bank code
              </label>
              <input
                class="chakra-input css-1hv7q3p"
                data-testid="bank-code-input"
                id="bank-code"
                placeholder="Bank code"
                type="text"
                value=""
              />
            </div>
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="chakra-form__label css-1qqx8g3"
                for="trx"
                id="trx-label"
              >
                Transaction ID
              </label>
              <input
                class="chakra-input css-1hv7q3p"
                data-testid="trx-id-input"
                id="trx"
                placeholder="Bank code"
                type="text"
                value=""
              />
            </div>
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="chakra-form__label css-1qqx8g3"
                for="business"
                id="business-label"
              >
                Business
              </label>
              <div
                class="chakra-select__wrapper css-42b2qy"
              >
                <select
                  class="chakra-select css-6jerdq"
                  data-testid="business-input"
                  id="business"
                >
                  <option
                    value=""
                  >
                    Select option
                  </option>
                </select>
                <div
                  class="chakra-select__icon-wrapper css-fmsrp"
                >
                  <svg
                    aria-hidden="true"
                    class="chakra-select__icon"
                    focusable="false"
                    role="presentation"
                    style="width: 1em; height: 1em; color: currentColor;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <button
              class="chakra-button css-pa74a0"
              data-testid="send-callback-button"
              type="button"
            >
              Send callback
            </button>
            <button
              class="chakra-button css-1g4dmh7"
              type="button"
            >
              Go back
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="chakra-container css-kvcvn0"
    >
      <div
        class="css-gmuwbf"
      >
        <h2
          class="chakra-heading css-zey6tx"
        >
          Send callback
        </h2>
      </div>
      <div
        class="css-rwokqt"
      >
        <div
          class="chakra-stack css-ynqzo5"
        >
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-1qqx8g3"
              for="va"
              id="va-label"
            >
              Virtual account
            </label>
            <input
              class="chakra-input css-1hv7q3p"
              data-testid="va-input"
              id="va"
              placeholder="Virtual Account"
              type="text"
              value=""
            />
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-1qqx8g3"
              for="bank-code"
              id="bank-code-label"
            >
              Bank code
            </label>
            <input
              class="chakra-input css-1hv7q3p"
              data-testid="bank-code-input"
              id="bank-code"
              placeholder="Bank code"
              type="text"
              value=""
            />
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-1qqx8g3"
              for="trx"
              id="trx-label"
            >
              Transaction ID
            </label>
            <input
              class="chakra-input css-1hv7q3p"
              data-testid="trx-id-input"
              id="trx"
              placeholder="Bank code"
              type="text"
              value=""
            />
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-1qqx8g3"
              for="business"
              id="business-label"
            >
              Business
            </label>
            <div
              class="chakra-select__wrapper css-42b2qy"
            >
              <select
                class="chakra-select css-6jerdq"
                data-testid="business-input"
                id="business"
              >
                <option
                  value=""
                >
                  Select option
                </option>
              </select>
              <div
                class="chakra-select__icon-wrapper css-fmsrp"
              >
                <svg
                  aria-hidden="true"
                  class="chakra-select__icon"
                  focusable="false"
                  role="presentation"
                  style="width: 1em; height: 1em; color: currentColor;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </div>
          <button
            class="chakra-button css-pa74a0"
            data-testid="send-callback-button"
            type="button"
          >
            Send callback
          </button>
          <button
            class="chakra-button css-1g4dmh7"
            type="button"
          >
            Go back
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
